### PR TITLE
Improve SubCompaction Partitioning

### PR DESCRIPTION
**Pull request by siying**
Summary:
Unit tests still haven't been fixed. Also need to add more tests. But I ran some simple fillrandom db_bench and the partitioning feels reasonable.

Test Plan: 
1. Make sure existing tests pass. This should cover some basic sub compaction logic to be correct and the partitioning result is reasonable;
2. Add a new unit test to ApproximateKeyAnchors()
3. Run some db_bench with max_subcompaction = 4 and watch the compaction is indeed partitioned evenly.

### CODE REVIEW

**Comment by rockeet** [comment]
This should be the implementation of #1842(Request: Optimization for subcompact: Add Get Random Keys)

File: table/block_based/block_based_table_reader.cc [hunk 1]
```diff
@@ -2045,6 +2045,53 @@ void BlockBasedTable::FullFilterKeysMayMatch(
...
+  IndexBlockIter iiter_on_stack;
+  auto iiter = NewIndexIterator(
+      read_options, /*disable_prefix_seek=*/false, &iiter_on_stack,
+      /*get_context=*/nullptr, /*lookup_context=*/nullptr);
+  std::unique_ptr<InternalIteratorBase<IndexValue>> iiter_unique_ptr;
+  if (iiter != &iiter_on_stack) {
+    iiter_unique_ptr.reset(iiter);
+  }
+
+  const uint64_t kMaxNumAnchors = uint64_t{128};
```
Comment (jay-zhuang) [inline table/block_based/block_based_table_reader.cc:2066]: Would it be better to make it proportional to the size of the file, so small files don't have to sample that much, and make sure it also works okay for huge files.

**Review by jay-zhuang — APPROVED** [review]
LGTM.

I think we should mention that it's a behavior change that sub-compaction number is more like be max out.

File: table/block_based/block_based_table_reader.cc (hunk 1 above)
Comment (siying) [reply table/block_based/block_based_table_reader.cc:2066]: I can add a comment. It's not easy to figure out inside the reader whether it is a smaller file or a larger file of the compaction.

**Pull request closed without merging**